File: packages/server/src/aggregate.ts
import { BUCKET_COUNT, emptyHistogram, mergeInto, percentileFromHistogram } from '@mo/shared';
import type { RollupRow } from './store/DuckStore.js';

export function histFrom(buckets: number[], counts: number[]): number[] {
  const h = emptyHistogram();
  for (let i = 0; i < buckets.length; i++) {
    const b = buckets[i];
    if (b >= 0 && b < BUCKET_COUNT) h[b] += counts[i] ?? 0;
  }
  return h;
}

export interface Agg {
  n: number;
  errors: number;
  sumDurNs: number;
  hist: number[];
}

export function newAgg(): Agg {
  return { n: 0, errors: 0, sumDurNs: 0, hist: emptyHistogram() };
}

export function addRow(a: Agg, r: RollupRow): void {
  a.n += r.n;
  a.errors += r.errors;
  a.sumDurNs += r.sum_dur_ns;
  mergeInto(a.hist, histFrom(r.buckets, r.counts));
}

export interface Summary {
  count: number;
  errorRate: number;
  throughputPerMin: number;
  avgMs: number;
  p50Ms: number;
  p95Ms: number;
  p99Ms: number;
}

export function summarize(a: Agg, windowMs: number): Summary {
  const minutes = Math.max(windowMs / 60_000, 1 / 60);
  return {
    count: a.n,
    errorRate: a.n ? a.errors / a.n : 0,
    throughputPerMin: a.n / minutes,
    avgMs: a.n ? a.sumDurNs / a.n / 1e6 : 0,
    p50Ms: percentileFromHistogram(a.hist, 0.5),
    p95Ms: percentileFromHistogram(a.hist, 0.95),
    p99Ms: percentileFromHistogram(a.hist, 0.99),
  };
}

/** Group rollup rows by a key and summarize each group. */
export function groupSummaries<K extends string>(
  rows: RollupRow[],
  keyOf: (r: RollupRow) => K | null,
  windowMs: number,
): Map<K, Summary & { agg: Agg }> {
  const groups = new Map<K, Agg>();
  for (const r of rows) {
    const k = keyOf(r);
    if (k === null) continue;
    let g = groups.get(k);
    if (!g) groups.set(k, (g = newAgg()));
    addRow(g, r);
  }
  const out = new Map<K, Summary & { agg: Agg }>();
  for (const [k, a] of groups) out.set(k, { ...summarize(a, windowMs), agg: a });
  return out;
}

/** Hourly time series of a metric for charting. */
export function timeseries(rows: RollupRow[], metric: 'throughput' | 'p95' | 'p50' | 'p99' | 'errorRate' | 'avg') {
  const byHour = new Map<number, Agg>();
  for (const r of rows) {
    let a = byHour.get(r.hour_ms);
    if (!a) byHour.set(r.hour_ms, (a = newAgg()));
    addRow(a, r);
  }
  return [...byHour.entries()]
    .sort((x, y) => x[0] - y[0])
    .map(([hour, a]) => {
      const s = summarize(a, 3_600_000);
      const value =
        metric === 'throughput'
          ? s.throughputPerMin
          : metric === 'errorRate'
            ? s.errorRate
            : metric === 'avg'
              ? s.avgMs
              : metric === 'p50'
                ? s.p50Ms
                : metric === 'p99'
                  ? s.p99Ms
                  : s.p95Ms;
      return { t: hour, value };
    });
}

/** Time-in-span-type breakdown (self-time approximation via span_type sums). */
export function breakdown(rows: RollupRow[]) {
  const byType = new Map<string, number>();
  for (const r of rows) byType.set(r.span_type, (byType.get(r.span_type) ?? 0) + r.sum_dur_ns);
  const total = [...byType.values()].reduce((a, b) => a + b, 0) || 1;
  return [...byType.entries()]
    .map(([type, sum]) => ({ type, ms: sum / 1e6, fraction: sum / total }))
    .sort((a, b) => b.ms - a.ms);
}

File: sqlite-server/aggregate.go
package main

import "sort"

type Agg struct {
	N        int64
	Errors   int64
	SumDurNs float64
	Hist     []float64
}

func histFrom(buckets, counts []int) []float64 {
	h := emptyHistogram()
	for i := 0; i < len(buckets); i++ {
		b := buckets[i]
		if b >= 0 && b < bucketCount && i < len(counts) {
			h[b] += float64(counts[i])
		}
	}
	return h
}

func newAgg() *Agg { return &Agg{Hist: emptyHistogram()} }

func (a *Agg) add(r RollupRow) {
	a.N += r.N
	a.Errors += r.Errors
	a.SumDurNs += r.SumDurNs
	mergeInto(a.Hist, histFrom(r.Buckets, r.Counts))
}

type Summary struct {
	Count            int64   `json:"count"`
	ErrorRate        float64 `json:"errorRate"`
	ThroughputPerMin float64 `json:"throughputPerMin"`
	AvgMs            float64 `json:"avgMs"`
	P50Ms            float64 `json:"p50Ms"`
	P95Ms            float64 `json:"p95Ms"`
	P99Ms            float64 `json:"p99Ms"`
}

func (a *Agg) summarize(windowMs int64) Summary {
	minutes := float64(windowMs) / 60000.0
	if minutes < 1.0/60 {
		minutes = 1.0 / 60
	}
	s := Summary{Count: a.N}
	if a.N > 0 {
		s.ErrorRate = float64(a.Errors) / float64(a.N)
		s.AvgMs = a.SumDurNs / float64(a.N) / 1e6
	}
	s.ThroughputPerMin = float64(a.N) / minutes
	s.P50Ms = percentileFromHistogram(a.Hist, 0.5)
	s.P95Ms = percentileFromHistogram(a.Hist, 0.95)
	s.P99Ms = percentileFromHistogram(a.Hist, 0.99)
	return s
}

// groupSummaries groups rollup rows by key and summarizes each group.
func groupSummaries(rows []RollupRow, keyOf func(RollupRow) (string, bool), windowMs int64) map[string]Summary {
	groups := map[string]*Agg{}
	for _, r := range rows {
		k, ok := keyOf(r)
		if !ok {
			continue
		}
		g := groups[k]
		if g == nil {
			g = newAgg()
			groups[k] = g
		}
		g.add(r)
	}
	out := map[string]Summary{}
	for k, a := range groups {
		out[k] = a.summarize(windowMs)
	}
	return out
}

type Point struct {
	T     int64   `json:"t"`
	Value float64 `json:"value"`
}

func timeseries(rows []RollupRow, metric string) []Point {
	byHour := map[int64]*Agg{}
	for _, r := range rows {
		a := byHour[r.HourMs]
		if a == nil {
			a = newAgg()
			byHour[r.HourMs] = a
		}
		a.add(r)
	}
	var hours []int64
	for h := range byHour {
		hours = append(hours, h)
	}
	sort.Slice(hours, func(i, j int) bool { return hours[i] < hours[j] })
	out := make([]Point, 0, len(hours))
	for _, h := range hours {
		s := byHour[h].summarize(3600000)
		var v float64
		switch metric {
		case "throughput":
			v = s.ThroughputPerMin
		case "errorRate":
			v = s.ErrorRate
		case "avg":
			v = s.AvgMs
		case "p50":
			v = s.P50Ms
		case "p99":
			v = s.P99Ms
		default:
			v = s.P95Ms
		}
		out = append(out, Point{T: h, Value: v})
	}
	return out
}

type BreakdownItem struct {
	Type     string  `json:"type"`
	Ms       float64 `json:"ms"`
	Fraction float64 `json:"fraction"`
}

func breakdownFrom(rows []RollupRow) []BreakdownItem {
	byType := map[string]float64{}
	total := 0.0
	for _, r := range rows {
		byType[r.SpanType] += r.SumDurNs
		total += r.SumDurNs
	}
	if total == 0 {
		total = 1
	}
	out := make([]BreakdownItem, 0, len(byType))
	for t, sum := range byType {
		out = append(out, BreakdownItem{Type: t, Ms: sum / 1e6, Fraction: sum / total})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Ms > out[j].Ms })
	return out
}
